### ISSUE TITLE
FEP-0009: Add new concurrency-related primitives to FreeST

### ISSUE THREAD

**Issue by bpalmeida** [freest-lang/freest3#173]
In GitLab by @glopes on Apr 27, 2023, 15:46

# FEP-0009: Add new concurrency-related primitives to FreeST
## Overview
  | Header             | Value             |
  | :----------------- | :---------------- |
  | **FEP**            | 0009              |
  | **Title**          | Add new concurrency-related primitives to FreeST              |
  | **Author**         | @glopes            |
  | **Status**         | Draft |
  | **Type**           | Enhancement              |
  | **Created**        | 27-april-2022              |
  | **FreeST-Version** | v3.0.0            |
  | **Depends on**      | -              |

## Abstract

As a concurrent programming language, having more control over threads and the overall concurrency system in FreeST would be helpful.

Haskell offers several abstractions so users can manipulate or be knowledgeable of the scheduler information in runtime. Most of their behaviour is currently inaccessible in FreeST, such as getting the ID of a thread or "blocking" a thread for a particular duration. 

## Motivation

Some languages are similar to FreeST in their approach and view towards concurrency. We frequently compare FreeST to the Go programming language to better explain and expose FreeST's advantages in what it is achieving. Both work with lightweight threads and channels as the (primary) source of communication between processes.

Other languages come to mind when considering constructing concurrent systems, such as Rust or even Java, as a more mainstream example. In the case of these languages, we (mostly) work with system threads, which are way more expensive and reflect a very different scheduler that manages their lifetime.

All these languages offer essential primitives that allow programmers to access information on their system (e.g. the number of cores available), ways of interacting with the runtime system or controlling concurrent units (threads). The primitives available vary accordingly with each language's needs and particular characteristics.

This behaviour is currently inaccessible in FreeST and is essential for implementing several concurrent patterns, complex structures and ambitious systems. Since we depend on Haskell's scheduler to manipulate and manage threads and are not allowed to interact with its runtime system, we can not implement the needed primitives for this behaviour.

## Specification

The following primitives I am proposing, borrowed from the Haskell `Control.Concurrent` library, are primitives I noticed to be meaningful and offer behaviour that FreeST does not currently allow:
- `getNumCores` - Return the number of system threads that can run truly simultaneously (on separate physical processors) at any given time.
  
  In concurrent programs, this primitive might be useful, for example, to control the number of processes that are "forked" programmatically or by offering the possibility of intuitively estimating it. 
  
  It is a small addition, but nearly every popular programming language focusing on concurrency allows this (Go and Rust, for example).

- `setNumCores` - Set the number of system threads that can run truly simultaneously (on separate physical processor at any given time). 
  
  Currently, FreeST programs are predefined to run on the maximum number of system threads possible in the project's `package.yaml` file in the `ghc-options: -main-is FreeST -threaded -rtsopts -with-rtsopts=-N` executable configuration line. Although, "It is strongly recommended that the number of capabilities is not set larger than the number of physical processor cores, and it may often be beneficial to leave one or more cores free to avoid contention with other processes in the machine.".
  
  Fortunately, it is possible to set the number of system threads to a different number by executing the program with the command `freest program.fst +RTS -Nx`, being *x* the number of system threads available to the program. However, a primitive that offers the possibility of altering this programmatically at any given moment might be an interesting addition.
  
  Once again, it is a small addition. However, this behaviour appears in programming languages that work with lightweight threads (Go, for example).

- `myThreadId` - Returns the Id of the calling thread.
- `threadCore` - Returns the number of the core on which the thread is currently running.
  
  This primitive would allow us to create structures that simulate the state of each thread system queue in run-time related to the distribution of lightweight threads in several system threads.
  
  For a personal application, I could use this to the ForkJoin granularity control technique called *Surplus*: Before creating a new task, the number of queued tasks (lightweight threads) in the current (system) thread that exceeds the number of tasks in other queues is compared to a parameterized threshold limit. If the surplus tasks count is higher than the threshold, the task will be executed sequentially. If the surplus tasks count is lower than the threshold, the task is created in parallel.

- `threadDelay` - Suspends the current thread for a given number of microseconds.
  
  This one is debatable. 
  Most distributed systems frequently use this primitive (or an equivalent one) to synchronise part of asynchronous systems in specific scenarios. There are a lot of legitimate real-world applications, such as avoiding busy waiting by implementing delays between attempts at some execution. Issue #161 proposes another application using this primitive.
  However, this mechanism is a critical source of deadlocks, something we try distancing ourselves from.
  
  Moreover, nearly every popular programming language focusing on distributed systems allows this (C, Go, and Rust, for example).

These primitives can be consulted at [Control.Concurrent](https://hackage.haskell.org/package/base-4.18.0.0/docs/Control-Concurrent.html) for more details.

## Syntax and Implementation

Most of the implementation should be straightforward. Like other Haskell primitives built into FreeST's Prelude, we have to add and adjust the mentioned primitives in the `Interpreter.Builtin`.

In the Prelude, we would maybe have the following signatures:
```Haskell
type ThreadId = ??

getNumCores : Int
setNumCores : Int -> ()

myThreadId : ThreadId
threadCore : ThreadId -> Int

threadDelay : Int -> ()
```

I left ThreadId as a to-be defined type because the original `myThreadId` and `threadCore` (called `threadCapability` in Haskell) return and receive a particular data type `ThreadId`: 
```haskell
data ThreadId = ThreadId ThreadID#
```
"A `ThreadId` is an abstract type representing a handle to a thread." and "(...) if you have a `ThreadId`, you essentially have a pointer to the thread itself."

We would probably need to give this data type some special treatment since it appears impossible to transform an Int or String into a Haskell `ThreadId`.

A more basic version of the primitives I am proposing could be:
```haskell
type ThreadId = String

myThreadId : ThreadId
threadCore : Int
```
`myThreadId` returns a conversion of the original Haskell `threadId` to a String ("The `Show` instance lets you convert an arbitrary-valued `ThreadId` to string form...").
As for `threadCore`, we could simplify its logic and behaviour by internally executing a `myThreadId`, returning the `ThreadId` of the current thread, and applying it to Haskell's `threadCapability` first parameter, returning the number of the capability on which the thread is currently running.

In short, the implementation of `getNumCores`, `setNumCores` and `threadDelay` is immediate. On the other hand, `myThreadId` and `threadCore` following my suggestion are also immediate; otherwise, it might require more time and effort for a more in-depth analysis for its implementation.

If FreeST happened to be written in another programming language, we would have to adjust the primitive association to a distinct language's concurrent API that interacts with a very particular scheduler. Then, FreeST would be significantly different, requiring a different approach to analyse its needs, which could impact this proposal.
